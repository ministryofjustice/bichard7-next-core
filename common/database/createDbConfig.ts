export type DbConfig = {
  host: string
  port: number
  username: string
  password: string
  max: number
  idle_timeout: number
  max_lifetime: number
}

const createDbConfig = (): DbConfig => ({
  host: process.env.DB_HOST ?? "localhost",
  port: Number(process.env.DB_PORT ?? "5432"),
  username: process.env.DB_USER ?? "bichard",
  password: process.env.DB_PASSWORD ?? "password",
  max: 10,
  idle_timeout: 20,
  max_lifetime: 60 * 30
})

export default createDbConfig
