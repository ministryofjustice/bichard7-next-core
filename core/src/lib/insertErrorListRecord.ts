import type { Sql } from "postgres"
import type { PromiseResult } from "src/comparison/types"
import convertAhoToXml from "src/serialise/ahoXml/generate"
import type { AnnotatedHearingOutcome, OrganisationUnitCodes } from "src/types/AnnotatedHearingOutcome"
import type ErrorListRecord from "src/types/ErrorListRecord"
import { QualityCheckStatus } from "src/types/ErrorListRecord"
import type ErrorListTriggerRecord from "src/types/ErrorListTriggerRecord"
import type { Phase1SuccessResult } from "src/types/Phase1Result"
import ResolutionStatus from "src/types/ResolutionStatus"

const generateDefendantName = (aho: AnnotatedHearingOutcome): string => {
  const personName = aho.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.DefendantDetail?.PersonName
  return `${personName?.FamilyName} ${personName?.GivenName}`
}

const generateErrorReport = (aho: AnnotatedHearingOutcome): string => {
  // Bichard generates this in the order they appear in the XML document so we need to do this too
  const ahoXml = convertAhoToXml(aho)
  const matches = ahoXml.matchAll(/<([^ |>]*)[^>]* Error="([^"]*)/gm)
  if (!matches) {
    return ""
  }
  // return ""
  return Array.from(matches)
    .map((m) => `${m[2]}||${m[1]}`)
    .join(", ")
}

const generateOrgForPoliceFilter = (forceOwner?: OrganisationUnitCodes): string => {
  if (!forceOwner?.SecondLevelCode) {
    return ""
  }
  if (forceOwner.ThirdLevelCode === null || forceOwner.ThirdLevelCode === "00") {
    return forceOwner.SecondLevelCode
  }

  return `${forceOwner.SecondLevelCode}${forceOwner.ThirdLevelCode}`
}

const convertResultToErrorListRecord = (result: Phase1SuccessResult): ErrorListRecord => {
  const generateFalseHasErrorAttributes = result.triggers.length > 0 && result.hearingOutcome.Exceptions.length === 0

  const caseElem = result.hearingOutcome.AnnotatedHearingOutcome.HearingOutcome.Case
  const hearing = result.hearingOutcome.AnnotatedHearingOutcome.HearingOutcome.Hearing
  const errorReport = generateErrorReport(result.hearingOutcome)
  return {
    message_id: hearing.SourceReference.UniqueID,
    phase: 1,
    error_status: result.hearingOutcome.Exceptions.length > 0 ? ResolutionStatus.UNRESOLVED : null,
    trigger_status: result.triggers.length > 0 ? ResolutionStatus.UNRESOLVED : null,
    error_quality_checked: result.hearingOutcome.Exceptions.length > 0 ? QualityCheckStatus.UNCHECKED : null,
    trigger_quality_checked: result.triggers.length > 0 ? QualityCheckStatus.UNCHECKED : null,
    defendant_name: generateDefendantName(result.hearingOutcome),
    trigger_count: result.triggers.length,
    is_urgent: caseElem.Urgent?.urgent ? 1 : 0,
    asn: caseElem.HearingDefendant.ArrestSummonsNumber,
    court_code: hearing.CourtHearingLocation.OrganisationUnitCode,
    annotated_msg: convertAhoToXml(result.hearingOutcome),
    updated_msg: convertAhoToXml(result.hearingOutcome, false, generateFalseHasErrorAttributes),
    error_report: errorReport,
    create_ts: new Date(),
    error_reason: errorReport.length > 0 ? errorReport.split("||")[0] : null,
    error_insert_ts: new Date(),
    trigger_insert_ts: new Date(),
    trigger_reason: result.triggers.length > 0 ? result.triggers[0].code : null,
    error_count: result.hearingOutcome.Exceptions.length,
    user_updated_flag: 0,
    msg_received_ts: new Date(),
    court_reference: caseElem.CourtReference.MagistratesCourtReference.slice(0, 11),
    court_date: hearing.DateOfHearing,
    ptiurn: caseElem.PTIURN,
    court_name: hearing.CourtHouseName,
    org_for_police_filter: generateOrgForPoliceFilter(caseElem.ForceOwner),
    court_room: hearing.CourtHearingLocation.BottomLevelCode || undefined,
    pnc_update_enabled: "Y"
  }
}

const insertErrorListRecord = async (sql: Sql, result: Phase1SuccessResult): PromiseResult<void> => {
  const errorListRecord = convertResultToErrorListRecord(result)
  try {
    const newRecordResult = await sql<ErrorListRecord[]>`
      INSERT INTO br7own.error_list ${sql(errorListRecord)} RETURNING *`

    if (!newRecordResult[0].error_id) {
      throw new Error("Error inserting to error_list table")
    }

    for (const trigger of result.triggers) {
      const triggerRecord: ErrorListTriggerRecord = {
        error_id: newRecordResult[0].error_id,
        trigger_code: trigger.code,
        status: ResolutionStatus.UNRESOLVED,
        create_ts: new Date()
      }
      if (trigger.offenceSequenceNumber) {
        triggerRecord.trigger_item_identity = String(trigger.offenceSequenceNumber)
      }
      await sql`
        INSERT INTO br7own.error_list_triggers ${sql(triggerRecord)}`
    }
  } catch (e) {
    const error = e as any
    if (error.severity !== "NOTICE") {
      console.error(e)
      return e as Error
    }
  }
}

export default insertErrorListRecord
