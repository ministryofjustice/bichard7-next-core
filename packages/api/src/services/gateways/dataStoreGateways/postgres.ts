import type { User } from "@moj-bichard7/common/types/User"
import postgresFactory from "../../db/postgresFactory"
import type DataStoreGateway from "../interfaces/dataStoreGateway"
import caseCanBeResubmitted from "./postgres/canCaseBeResubmitted"
import fetchUserByUsername from "./postgres/fetchUserByUsername"

/*
 * We don't seem to have any logic for updating the AHO in core. It is all in the UI
 *
 * TODO: Check user's feature flags
 * TODO: UI `courtCase.canResolveOrSubmit(user)` only permission Exceptions so do we care about triggers?
 * TODO: Check the User doesn't have excluded trigger codes
 * TODO: Check what permission the user has (triggers or exceptions) and add lock
 * TODO: If case needs to be locked add AuditLogEvent
 *       Should be handled by court-cases/:caseId/index
 * TODO: amend the court case
 *       Shouldn't auto save done this?
 *       The UI creates a force owner and edits the aho if one doesn't exist
 *       Applies amendments to the aho
 *       Only updates if Phase 1
 *       Inserts notes
 *       Returns updated AHO
 */
class Postgres implements DataStoreGateway {
  protected readonly db = postgresFactory()

  async fetchUserByUsername(username: string): Promise<User> {
    return await fetchUserByUsername(this.db, username)
  }

  async canCaseBeResubmitted(username: string, caseId: number, forceIds: number[]): Promise<boolean> {
    return await caseCanBeResubmitted(this.db, username, caseId, forceIds)
  }
}

export default Postgres
