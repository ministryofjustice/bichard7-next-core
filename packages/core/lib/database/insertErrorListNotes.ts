import type { PostgresError, Sql } from "postgres"
import type ErrorListNoteRecord from "../../types/ErrorListNoteRecord"

const insertErrorListNotes = async (db: Sql, error_id: number, notes: (string | null)[]): Promise<void> => {
  try {
    for (const note of notes) {
      if (!note) {
        continue
      }

      const noteRecord: ErrorListNoteRecord = {
        error_id,
        note_text: note,
        user_id: "System",
        create_ts: new Date()
      }
      await db`
        INSERT INTO br7own.error_list_notes ${db(noteRecord)}`
    }
  } catch (e) {
    const error = e as PostgresError
    if (error.severity !== "NOTICE") {
      console.error(error)
      throw error
    }
  }
}

export default insertErrorListNotes
