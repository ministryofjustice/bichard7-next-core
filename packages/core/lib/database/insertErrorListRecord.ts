import type { PostgresError, Sql } from "postgres"
import type ErrorListRecord from "../../types/ErrorListRecord"
import type PhaseResult from "../../types/PhaseResult"
import convertResultToErrorListRecord from "./convertResultToErrorListRecord"

const insertErrorListRecord = async (db: Sql, result: PhaseResult): Promise<number> => {
  const errorListRecord = convertResultToErrorListRecord(result)
  let newRecordId: number | undefined
  try {
    const newRecordResult = await db<ErrorListRecord[]>`
      INSERT INTO br7own.error_list ${db(errorListRecord)} RETURNING *`

    if (!newRecordResult[0].error_id) {
      throw new Error("Error inserting to error_list table")
    }

    newRecordId = newRecordResult[0].error_id
  } catch (e) {
    const error = e as PostgresError
    if (error.severity !== "NOTICE") {
      console.error(e)
      throw error
    }
  }

  if (newRecordId === undefined) {
    throw new Error("Error inserting new error_list record")
  }

  return newRecordId
}

export default insertErrorListRecord
