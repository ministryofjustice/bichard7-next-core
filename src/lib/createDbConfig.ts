export type DbConfig = {
  host: string
  port: number
  username: string
  password: string
}

const createDbConfig = (): DbConfig => ({
  host: process.env.DB_HOST ?? "localhost",
  port: Number(process.env.DB_PORT ?? "5434"),
  username: process.env.DB_USER ?? "bichard",
  password: process.env.DB_PASSWORD ?? "password"
})

export default createDbConfig
