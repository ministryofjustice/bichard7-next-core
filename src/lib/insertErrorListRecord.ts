import type { Client } from "pg"
import type { PromiseResult } from "src/comparison/types"
import convertAhoToXml from "src/serialise/ahoXml/generate"
import type { AnnotatedHearingOutcome } from "src/types/AnnotatedHearingOutcome"
import type { Phase1SuccessResult } from "src/types/Phase1Result"

const generateDefendantName = (aho: AnnotatedHearingOutcome): string => {
  const personName = aho.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.DefendantDetail?.PersonName
  return `${personName?.FamilyName} ${personName?.GivenName}`
}

const insertErrorListRecord = async (db: Client, record: Phase1SuccessResult): PromiseResult<void> => {
  try {
    const caseElem = record.hearingOutcome.AnnotatedHearingOutcome.HearingOutcome.Case
    const hearing = record.hearingOutcome.AnnotatedHearingOutcome.HearingOutcome.Hearing
    const newRecordResult = await db.query(
      "INSERT INTO br7own.error_list (\
        message_id, defendant_name, phase, trigger_count, is_urgent, annotated_msg, error_report,\
        create_ts, error_count, user_updated_flag, msg_received_ts, court_reference, court_date, ptiurn,\
        court_name, org_for_police_filter, court_room)\
        VALUES ($1, $2, 1, $3, $4, $5, $6, NOW(), $7, 0, $8, $9, $10, $11, $12, $13, $14) RETURNING *",
      [
        hearing.SourceReference.UniqueID,
        generateDefendantName(record.hearingOutcome),
        record.triggers.length,
        0,
        convertAhoToXml(record.hearingOutcome),
        "",
        record.hearingOutcome.Exceptions.length,
        new Date(),
        caseElem.CourtReference.MagistratesCourtReference,
        hearing.DateOfHearing,
        caseElem.PTIURN,
        hearing.CourtHouseName,
        `${caseElem.ForceOwner?.SecondLevelCode}${caseElem.ForceOwner?.ThirdLevelCode}  `,
        hearing.CourtHearingLocation.BottomLevelCode
      ]
    )

    for (const trigger of record.triggers) {
      await db.query(
        "INSERT INTO br7own.error_list_triggers (error_id, trigger_code, trigger_item_identity, status, create_ts) VALUES ($1, $2, $3, 0, NOW())",
        [newRecordResult.rows[0].error_id, trigger.code, trigger.offenceSequenceNumber]
      )
    }
  } catch (e) {
    console.error(e)
    return e as Error
  }
}

export default insertErrorListRecord
