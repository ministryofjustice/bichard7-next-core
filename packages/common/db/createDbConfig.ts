import postgres from "postgres"

export type DbConfig = {
  host: string
  port: number
  username: string
  password: string
  max: number
  idle_timeout: number
  max_lifetime: number
  ssl: boolean | { rejectUnauthorized: false }
  types: Record<string, postgres.PostgresType>
}

const createDbConfig = (): DbConfig => ({
  host: process.env.DB_HOST ?? "localhost",
  port: Number(process.env.DB_PORT ?? "5432"),
  username: process.env.DB_USER ?? "bichard",
  password: process.env.DB_PASSWORD ?? "password",
  ssl: process.env.DB_SSL === "true" ? { rejectUnauthorized: false } : false,
  max: 10,
  idle_timeout: 20,
  max_lifetime: 60 * 30,
  types: {
    bigint: postgres.BigInt
  }
})

export default createDbConfig
