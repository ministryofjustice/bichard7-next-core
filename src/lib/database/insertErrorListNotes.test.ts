import postgres from "postgres"
import type ErrorListNoteRecord from "src/types/ErrorListNoteRecord"
import generateMockPhase1Result from "tests/helpers/generateMockPhase1Result"
import createDbConfig from "../createDbConfig"
import insertErrorListNotes from "./insertErrorListNotes"
import insertErrorListRecord from "./insertErrorListRecord"

const dbConfig = createDbConfig()
const db = postgres({
  ...dbConfig,
  types: {
    date: {
      to: 25,
      from: [1082],
      serialize: (x: string): string => x,
      parse: (x: string): Date => {
        return new Date(x)
      }
    }
  }
})

const snapshotExclusions = {
  create_ts: expect.any(Date),
  error_id: expect.any(Number),
  note_id: expect.any(Number)
}

describe("insertErrorListNotes", () => {
  beforeEach(async () => {
    await db`DELETE FROM br7own.error_list`
    await db`DELETE FROM br7own.error_list_notes`
  })

  it("should insert the error list notes", async () => {
    const notes = ["Trigger codes: 1 x TRPR0001.", "Error codes: 1 x HO100100."]
    const phase1Result = generateMockPhase1Result()

    const recordId = await insertErrorListRecord(db, phase1Result)
    await insertErrorListNotes(db, recordId, notes)

    const insertedRecords = await db<ErrorListNoteRecord[]>`
      SELECT * FROM br7own.error_list_notes WHERE error_id = ${recordId}`

    expect(insertedRecords).toHaveLength(2)
    expect(insertedRecords[0]).toMatchSnapshot(snapshotExclusions)
    expect(insertedRecords[1]).toMatchSnapshot(snapshotExclusions)
  })
})
