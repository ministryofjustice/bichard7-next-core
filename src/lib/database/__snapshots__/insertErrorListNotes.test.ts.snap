// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertErrorListNotes should insert the error list notes 1`] = `
{
  "create_ts": Any<Date>,
  "error_id": Any<Number>,
  "note_id": Any<Number>,
  "note_text": "Trigger codes: 1 x TRPR0001.",
  "user_id": "System",
}
`;

exports[`insertErrorListNotes should insert the error list notes 2`] = `
{
  "create_ts": Any<Date>,
  "error_id": Any<Number>,
  "note_id": Any<Number>,
  "note_text": "Error codes: 1 x HO100100.",
  "user_id": "System",
}
`;
